### DISCUSSION TITLE
Is Nginx recommended for efficient deployment?

### DISCUSSION THREAD

**Discussion by ktobah** [fastapi/fastapi#3696] [Questions] [answered]
Hello,

I am deploying my app using the official docker image that has uvicorn and gunicorn. I was wondering if using Nginx with them will be efficient?  Will gunicorn be enough to handle a load of requests or should this be left to Nginx?

Thank you.

**Comment by Atem18** [comment]
@ktobah Gunicorn/Uvicorn are used to answer "Python calls", meaning when there is Python code to be executed. If you have static assets to serve like CSS/JS or pure HTML, it should be served by Nginx.

**Reply by ktobah** [reply]
@Atem18 thank you for your answer.

My concern is more about requests handling. Will the unicorn and uvicorn combination be enough to handle a large number of requests from clients or should Nginx be added to the mix for that? Given that the major thing there is python code to execute. Very few or inexisting static files or HTML to serve.

**Comment by StephenPCG** [comment, marked as answer]
@ktobah You should always use nginx if you have slow clients. Slow clients means clients from Internet (not from other servers in the same cluster).

An http server has to keep connection context in memory until the connection is finished and disconnected. In the following scenary:

```
python (fastapi, uvicorn, gunicorn) <--> nginx <--> client
```

Nginx is just a proxy, it requires very few memory (usually far less than 1MB) for each connection, so 1GB memory is enough for more than 1000 concurrent connections.

Python web server, on the other hand, has to keep more data in memory (usually much more than 1MB)  for each connection.

Let's imagine we have an API endpoint, the python web server needs 0.1s to process the request, but the client is slow (small bandwidth, or high latency), needs 1s to finish receiving all data and close the connection.

```mermaid
sequenceDiagram
    client ->>+ nginx: start request
    nginx ->>+ python: proxy send request
    note right of python: this session only lives for 0.1s
    python ->>- nginx: send response in 0.1s, and finish this connection
    note left of nginx: this session lasts for 1s
    nginx ->>- client: nginx buffers response (in disk if large) and serve to client in 1s.
```

If there is no nginx, then the python web server needs to keep the session for 1s, which wastes a lot of memory.

So, in most cases, no matter what language/framework does the backend server use, it is always recommended to put an nginx in front, unless the backend server only serves fast clients, e.g. other servers in the same cluster. Nginx is better at dealing with slow clients.

**Reply by ktobah** [reply]
Hi @StephenPCG 

Thank you for this explanation. Now, things are clear for me.
